lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,802 @@
+
+  # No authorization cookie again
+  $t->get_ok('/')
+    ->status_is(401)
+    ->content_is('Please log in');
+
+We can also inspect cookies in responses for special values through the
+transaction's response (L<Mojo::Message::Response>) object:
+
+  $t->get_ok('/');
+
+  my ($cookie) = grep { $_->name eq 'smarty' } @{$t->tx->res->cookies};
+  like($cookie && $cookie->to_string, qr(smarty=pants), 'cookie found');
+
+=head2 Custom transactions
+
+Let's say we have an application that responds to a new HTTP verb C<RING> and
+to use it we must also pass in a secret cookie value. This is not a
+problem. We can test the application by creating a L<Mojo::Transaction>
+object, setting the cookie, then passing the transaction object to
+C<request_ok>:
+
+  # Use custom "RING" verb
+  my $tx = $t->ua->build_tx(RING => '/doorbell');
+
+  # Set a special cookie
+  $tx->req->cookies({Secret => "don't tell anybody"});
+
+  # Make the request
+  $t->request_ok($tx)
+    ->status_is(200)
+    ->json_is('/status' => 'ding dong');
+
+=head2 Testing WebSocket web services
+
+While the message flow on WebSocket connections can be rather dynamic, it more
+often than not is quite predictable, which allows this rather pleasant
+L<Test::Mojo> API to be used.
+
+  use Mojo::Base -strict;
+
+  use Test::More;
+  use Test::Mojo;
+
+  # Include application
+  use FindBin;
+  require "$FindBin::Bin/../echo.pl";
+
+  # Test echo web service
+  my $t = Test::Mojo->new;
+  $t->websocket_ok('/echo')
+    ->send_ok('Hello Mojo!')
+    ->message_ok
+    ->message_is('echo: Hello Mojo!')
+    ->finish_ok;
+
+  # Test JSON web service
+  $t->websocket_ok('/echo.json')
+    ->send_ok({json => {test => [1, 2, 3]}})
+    ->message_ok
+    ->json_message_is('/test', [1, 2, 3])
+    ->finish_ok;
+
+  done_testing();
+
+=head2 Extending L<Test::Mojo>
+
+If you see that you're writing a lot of test assertions that aren't chainable,
+you may benefit from writing your own test assertion. This example comes from
+the L<Test::Mojo> documentation, but is worth examining closer:
+
+  my $location_is = sub {
+    my ($t, $value, $desc) = @_;
+    $desc ||= "Location: $value";
+    local $Test::Builder::Level = $Test::Builder::Level + 1;
+    return $t->success(is($t->tx->res->headers->location, $value, $desc));
+  };
+  $t->get_ok('/')
+    ->status_is(302)
+    ->$location_is('http://mojolicious.org')
+    ->or(sub { diag 'Must have been Joel!' });
+
+We create an anonymous subroutine (C<$location_is>) whose method signature
+matches other C<*_is> methods in L<Test::Mojo>. We create a default test
+description (C<$desc>), set the L<Test::Builder> C<Level> global variable one
+level higher (which tells L<Test::Builder> how far up the call stack to look
+when something fails). Finally we use L<Test::More>'s C<is> method to compare
+the location header with the expected header value. We wrap that in the
+C<success> method, which propagates the object for method chaining.
+
+=head3 Subclassing
```

Comment:
Since `Test::Mojo::WithRoles` is a third party module i'm not sure it should get a section in this guide. I'll have to discuss this with the rest of the team. If the `Test::Mojo::Role` namespace wasn't blocked i might just have added the functionality to `Test::Mojo`, since it's the right way to deal with this problem.